Terminology

kvyh opened the pull request:
Added a .rst page for the terminology "agreed" upon during the python in astronomy conference.

bsipocz commented:
Maybe it's worth using different font formatting for the phrases?

bmorris3 commented:
@kvyh I don't think you meant to do this, but in the future you should allow the tests to run when you make docs updates. There are tests to make sure the docs build successfully.

bmorris3 commented on docs/Terminology.rst:
```diff
@@ -0,0 +1,25 @@
+==================
+Determined during the python in astronomy conference:
+
+* **observing block** (OB): the minimum schedulable block
```
There's [a case to be made](https://github.com/astropy/astroplan/pull/151#issuecomment-224250861) that the observing block is distinct from a minimum schedulable block. I think OB needs a more generic definition, like maybe "a single or group of visits to one or multiple targets, with associated instrument configurations and observational constraints."

bmorris3 commented on docs/Terminology.rst:
```diff
@@ -0,0 +1,25 @@
+Determined during the python in astronomy conference:
+
+* **observing block** (OB): the minimum schedulable block
+* **priority**: number assigned to the **observing block** by the user (or TAC) that 
```
@ejeschke Is this how we settled on using "priority"? My intuition is that "priority" is user-assigned and "rank" is TAC-assigned.

ejeschke replied:
I think it is a good use of the term--that's how we define it at Subaru.  The user can provide priorities to help inform the scheduler of the desirability of one OB vs another if both are being considered for the same slot.  See my comment below for how we define TAC score.

bmorris3 commented on docs/Terminology.rst:
```diff
@@ -0,0 +1,25 @@
+  **observing blocks** and produces a schedule
+* **Weights** (“user defined”?): preferences for which constraint's **scores** matter most 
+  (i.e. I care more about getting dark time than getting a low airmass)
+  
```
Additional terms to consider: 
* rank, i.e. a "priority" assigned by a TAC?
* visit (I _think_ this one is well defined). This would distinguish a single visit from an observing block, since a visit may be its own OB or a component of an OB with multiple visits that must be scheduled together.
* names for various schedulers, i.e. the "greedy scheduler", which different people at the conference thought meant different things

eteq commented on docs/faq/terminology.rst:
```diff
@@ -0,0 +1,33 @@
+***************
+
+Scheduling
+==================
```
Super minor, but it's generally best-practice to have as many `=` characters as the text just above.  So thie would be:
```
Scheduling
==========
```

and similar for the `*`s above (around "Terminology")

eteq commented on docs/faq/terminology.rst:
```diff
@@ -0,0 +1,33 @@
+
+Schedulers (not yet implemented):
+* **Sequential**: starts from the beginning of the time range and schedules the OB
+   with the best **score**, that hasn't already been scheduled, for that time.
```
This has one space too much of indentation, and that's one of the sources of the warnings

eteq commented on docs/faq/terminology.rst:
```diff
@@ -0,0 +1,33 @@
+  (i.e. I care more about getting dark time than getting a low airmass)
+
+Schedulers (not yet implemented):
+* **Sequential**: starts from the beginning of the time range and schedules the OB
```
I think you need whitespace above this line for it to correctly render in the final docs

eteq commented on docs/faq/terminology.rst:
```diff
@@ -0,0 +1,33 @@
+
+Scheduling
+==================
+Determined during the python in astronomy conference:
```
You should add just a bit more background here.  Something along the lines of:

Scheduling observations is a task with a lot of history but not a lot of formal terminology.  Here we define the terminology used in astroplan for scheduling-related tasks.  None of this is meant to be particularly official in any manner other than we try to use these terms consistently in the code and documentation of astroplan.

eteq commented:
One other suggestion, @kvyh, in case you don't know:  You should try doing ``python setup.py build_sphinx`` when you make changes that affect the documentation.  If you do that, the docs will build *locally* for you, and you can see if there are errors/warnings, and then you can fix them. 

Even if you don't do that, using github you can see those results -  the little red "x" are for commits that these didn't build correctly, and then if you follow the "travis-ci" links, and click on the sub-builds with their own red x, you'll get a build log that shows the sphinx build log, and from that you can look for error messages that caused the problem.

Hope that helps!

eteq commented on docs/faq/terminology.rst:
```diff
@@ -0,0 +1,43 @@
+  (i.e. I care more about getting dark time than getting a low airmass)
+
+
+The different schedulers that are already implemented (none are implemented yet) 
```
This should be rephrased a bit - just say straight-up what the present situation is - that these are not implemented, and this is a *tentative* set of possible schedulers.  Right now it doesn't make sense because it says "that are already implemented (none are implemented yet)".

Sorry if this seems picky, but in my experience it's best to be as accurate as possible for documentation because it tends to not get updated as promptly as code.  And users often get *very* confused if the documentation doesn't match the code.

eteq commented:
Left one more comment, but other than that this looks good, @kvyh.

(One other tip, @kvyh - it's usually good to leave a comment when you push up new changes.  Otherwise it's not clear whether you're done making changes, or if you're still working on it. Clear communication with *everyone* helps these sorts of projects a lot.)

wtgee commented:
Small thing: TAC (listed under **rank**) should be spelled out the first time or defined as a separate entry.

wtgee commented:
Also, the **score** is defined as being either 0 or 1 but the **weights** are not specified. Can the weighting factor be defined *between* 0 and 1?

kvyh commented:
I moved the description of score onto the score and changed the descriptions to better match the current state of astroplan. I also added the range of weights supported.

bmorris3 commented:
@kvyh A proposal like an observing proposal? I think that's a universal word.

eteq commented on docs/faq/terminology.rst:
```diff
@@ -0,0 +1,46 @@
+The terms used are as follows:
+
+* **visit**: a period of time spent observing a single target
+* **observing block** (OB): the minimum schedulable block (tentative: a group of visits)
```
I don't quite understand ``(tentative: a group of visits)``?  A would think an OB is always at least a visit or *smaller*: for example, you might want to observe the same target in two different filters - in that case there are two OBs but just one visit

bsipocz replied:
Just a comment from the sidelines: I'm surprised that there isn't a separate MSB in addition to an OB. I think in Erik's example above would work as two separate OBs (due to the filter difference), or one big one. I think it's clearer to have the first and join them into an MSB, but one can argue the other way around.
However consider the case when you do a photometric survey when one data point is no data point, so you want to have the exact same OB repeated over a certain number of times to form an MSB.

kvyh replied:
The way that it seems we are going with the definitions, is for an `ObservingBlock` to be the MSB (a specific target, duration and instrument configuration) which could then be repeated if necessary or joined with other `ObservingBlocks` into `BlockGroups`.

kvyh replied:
It is also possible that I have absolutely no idea what you mean by an MSB, in which case it would be good to have definition for it.

eteq commented on docs/faq/terminology.rst:
```diff
@@ -0,0 +1,46 @@
+* **priority**: number assigned to the **observing block** by the user that 
+  defines its precedence within the set of blocks to be scheduled. Should probably
+  also be on a 0->1 (0=no good, 1=good) scale, or rescaled within the scheduler
+* **rank**: like a **priority**, but defined by a time allocation committee (TAC) rather
```
I wouldn't say this *has* to be by a TAC - maybe instead "defined by an external agency like a time allocation committee (TAC)".  Also, I think this is a bit problematic because it's assigned to an entire "proposal" which isn't really defined here.  I agree with @bmorris3 that "proposal" is a well-understood term in general, but in the context of this definition, it's not clear.

So perhaps this "rank" idea needs to be re-thought as something more generic roughly like "a priority that applies to a specific set of observing blocks, often set by an agent other than the observer".  Perhaps give the TAC & proposal as a specific *example* of this, but by phrasing it as relative to a set of OBs, it's clearer how it fits in with this framework.

eteq commented on docs/faq/terminology.rst:
```diff
@@ -0,0 +1,46 @@
+  than a user and applied to an entire proposal.
+* **constraint**: sets limits and can yield **scores** based on inputs. Currently
+  only boolean scores are implemented in the code.
+* **score**: the value returned by evaluating a **constraint**. Can be 
```
I think you want this to be "a constraint or multiple constraints" - when discussing the scorekeeper, ther terminology " a cumulative score to the OB" is used, which means this word can also apply to an OB.  I think that makes sense and is just fine, but you should note in the definition that a score can be cumulative on an OB in addition to being created by a constraint

eteq commented:
For some reason I thought this had been merged already... but apparently not.  At any rate, @kvyh, I left a few comments above that I think are pretty straightforward to fix, and then I think this is all set to merge!

kvyh commented:
Ok, I have updated a number of the definitions in light of the conversation here.

bmorris3 commented:
@kvyh I'm working on re-running tests on the PRs that should fix the IERS errors and such. If this PR passes now, I'll merge.

bsipocz commented:
@bmorris3 - This should be rebased, otherwise those changes are invisible.

bmorris3 commented:
@bsipocz The tests pass now anyway? I think merging #177 did the trick.
